core/example/tutorial/13_halo_exchange/halo_exchange_example.cpp

```diff
@@ -0,0 +1,168 @@
+void haloExchangeExample()
+{
+    /*
+      The halo is a communication plan designed from halo exchange where some
+      locally-owned elements on each rank are used as ghost data on other
+      ranks. The halo supplies both forward and reverse communication
+      operations. In the forward operation (the gather), data is sent from the
+      uniquely-owned decomposition to the ghosted decomposition. In the
+      reverse operation (the scatter), data is sent from the ghosted
+      decomposition back to the uniquely-owned decomposition and collisions
+      are resolved.
+
+      In this example we will demonstrate building a halo communication
+      plan and performing both scatter and gather operations.
+
+      Note: The halo uses MPI for data movement. MPI is initialized
+      and finalized in the main function below.
+
+      Note: The halo uses GPU-aware MPI communication. If AoSoA data is
+      allocated in GPU memory, this feature will be used automatically.
+    */
+
+    /*
+       Get parameters from the communicator. We will use MPI_COMM_WORLD for
+       this example but any MPI communicator may be used.
+    */
+    int comm_rank = -1;
+    MPI_Comm_rank( MPI_COMM_WORLD, &comm_rank );
+    int comm_size = -1;
+    MPI_Comm_size( MPI_COMM_WORLD, &comm_size );
+
+    /*
+      Declare the AoSoA parameters.
+    */
+    using DataTypes = Cabana::MemberTypes<double,double>;
+    const int VectorLength = 8;
+    using MemorySpace = Kokkos::HostSpace;
+
+    /*
+       Create the AoSoA.
+    */
+    int num_tuple = 100;
+    Cabana::AoSoA<DataTypes,MemorySpace,VectorLength> aosoa( num_tuple );
+
+    /*
+      Create slices and assign data.
+     */
+    auto slice_0 = aosoa.slice<0>();
+    auto slice_1 = aosoa.slice<1>();
+    for ( int i = 0; i < num_tuple; ++i )
+    {
+        slice_0(i) = 1.0;
+        slice_1(i) = 1.0;
+    }
+
+    /*
+      Build a halo where the last 10 elements are sent to the next rank.
+    */
+    int local_num_send = 10;
+    Kokkos::View<int*,MemorySpace> export_ranks( "export_ranks", local_num_send );
+    Kokkos::View<int*,MemorySpace> export_ids( "export_ids", local_num_send );
+
+    // Last 10 elements (elements 90-99) go to the next rank. Note that this
+    // view will most often be filled within a parallel_for but we do so in
+    // serial here for demonstration purposes.
+    int next_rank = ( comm_rank == comm_size - 1 ) ? 0 : comm_rank + 1;
+    for ( int i = 0; i < local_num_send; ++i )
+    {
+        export_ranks(i) = next_rank;
+        export_ids(i) = i + num_tuple - 10;
+    }
+
+    /*
+      We have two ways to make a halo. In the first case we know where
+      we are sending the data to but not who we are receiving data from. In
+      the second we know the topology of the communication plan (i.e. who we
+      send and receive from). We know that we will only send/receive from this
+      rank and the next rank so use that information in this case.
```

Comment:
State what is the motivation for doing so (doing less work when constructing `halo`)

Reply:
added a comment discussing the extra global communication needed when the neighbor list is not supplied